src/main/groovy/pl/allegro/tech/build/axion/release/infrastructure/git/GitRepository.groovy

```diff
@@ -90,26 +99,36 @@ class GitRepository implements ScmRepository {
     }
 
     @Override
+    void dropTag(String tagName) {
+        repository.tag.remove(names: [tagName])
+    }
+
+    @Override
     void push(ScmIdentity identity, ScmPushOptions pushOptions) {
         push(identity, pushOptions, false)
     }
 
     void push(ScmIdentity identity, ScmPushOptions pushOptions, boolean all) {
-        identity.useDefault ? callPush(pushOptions, all) : callLowLevelPush(identity, pushOptions, all)
+        callPush(identity, pushOptions, all)
     }
 
-    private void callPush(ScmPushOptions pushOptions, boolean all) {
+    private void callPush(ScmIdentity identity, ScmPushOptions pushOptions, boolean all) {
         if (!pushOptions.pushTagsOnly) {
-            repository.push(remote: pushOptions.remote, all: all)
+            def regularPush = pushCommand(identity, pushOptions.remote, all)
+            verifyPushResults(regularPush.call())
         }
-        repository.push(remote: pushOptions.remote, tags: true, all: all)
+
+        def tagsPush = pushCommand(identity, pushOptions.remote, all).setPushTags()
+        verifyPushResults(tagsPush.call())
     }
 
-    private void callLowLevelPush(ScmIdentity identity, ScmPushOptions pushOptions, boolean all) {
-        if (!pushOptions.pushTagsOnly) {
-            pushCommand(identity, pushOptions.remote, all).call()
-        }
-        pushCommand(identity, pushOptions.remote, all).setPushTags().call()
+    private void verifyPushResults(Iterable<PushResult> pushResults) {
+        PushResult pushResult = ++pushResults.iterator()
+        Iterator<RemoteRefUpdate> remoteUpdates = pushResult.getRemoteUpdates().iterator()
+
+        remoteUpdates
+            .find { it.getStatus() == RemoteRefUpdate.Status.REJECTED_OTHER_REASON }
```

Comment:
i would go for `it.getStatus != SUCCESS`, so we log any errors (including rejection because of ahead of remote etc)

Reply:
I've also considered this, but there are some problems with this approach. There is a RemoteRefUpdate.Status.OK that we could use in place of suggested "SUCCESS", but:

1. UP_TO_DATE status is not really an error, so us failing on it doesn't sound right.
2. I think that applying the "! = OK" we actually lose a bit of control about what we are actually checking here. Even if we add UP_TO_DATE, in rather unlikely scenario when some new version of the library actually adds new "non-negative" status here, that check could lead to a lot of frustration.

But I find your suggestion rather pragmatic so I'll just go on and do it